File: LinkedList/PHP/DoublyLinkedList.php
<?php
/**
 * Created by PhpStorm.
 * User: saeed
 * Date: 24/11/16
 * Time: 05:16 AM
 */

namespace LinkedList;
require_once "DoublyLinkedListNode.php";

class DoublyLinkedList
{
    private $head = null;
    private $tail = null;
    private $count = 0;

    public function length(){
        return $this->count;
    }
    //Get
    //O(1)
    public function getHead()
    {
        if ($this->head == null)
        {
            return $this->head;
        }
        else
        {
            return $this->head->value;
        }
    }
    //O(1)
    public function getTail()
    {
        if ($this->tail == null)
        {
            return $this->tail;
        }
        else
        {
            return $this->tail->value;
        }
    }
    //O(n)
    public function get($index)
    {
        if ($index < 0 || $index > $this->length()){
            return null;
        }
        else{
            if ($index == 0){
                return $this->getHead();
            }
            else{
                if ($index == $this->length()){
                    return $this->getTail();
                }
                else{
                    $cur = $this->head;
                    $c =0;
                    while ($cur != null){
                        if ($index == $c){
                            return $cur->value;
                        }
                        $c++;
                        $cur = $cur->next;
                    }
                }
            }
        }
    }

    //Add
    //O(1)
    public function addFirst($value)
    {
        if ($this->head == null){
            $node = new DoublyLinkedListNode($value);
            $this->head = $node;
            $this->tail = $node;
        }
        else{
            $node = new DoublyLinkedListNode($value);
            $this->head->previous = $node;
            $node->next = $this->head;
            $this->head = $node;
        }
        $this->count--;
    }
    //O(1)
    public function addLast($value)
    {
        if ($this->tail == null){
            $node = new DoublyLinkedListNode($value);
            $this->head = $node;
            $this->tail = $node;
        }
        else{
            $node = new DoublyLinkedListNode($value);
            $this->tail->next = $node;
            $node->previous = $this->tail;
            $this->tail = $node;
        }
        $this->count++;
    }
    //O(n)
    public function add($value, $index)
    {
        if ($index < 0 || $index > $this->length()){
            // throw new \Exception();
        }
        else{
            if ($index == 0){
                return $this->addFirst($value);
            }
            else{
                if ($index == $this->length()){
                    return $this->addLast($value);
                }
                else{
                    $cur = $this->head;
                    $c =0;
                    while ($cur != null){
                        if ($index == $c){
                            $node = new DoublyLinkedListNode($value);
                            $node->previous = $cur;
                            $node->next = $cur->next;
                            $cur->next->previous = $node;
                            $this->count++;
                        }
                        $c++;
                        $cur = $cur->next;
                    }
                }
            }
        }
    }

    //Get
    //O(1)
    public function removeFirst()
    {
        if ($this->head != null){
            $temp = $this->head;
            $this->head = $this->head->next;
            $temp->next =  null;
            $this->head->previous = null;
            $this->count--;
        }
    }
    //O(1)
    public function removeLast()
    {
        if ($this->tail != null){
            $temp = $this->tail;
            $this->tail = $this->tail->previous;
            $temp->next =  null;
            $this->tail->next = null;
            $this->count--;
        }
    }
    //O(n)
    public function remove($index)
    {
        if ($index < 0 || $index > $this->length()){
            // throw new \Exception();
        }
        else{
            if ($index == 0){
                return $this->removeFirst();
            }
            else{
                if ($index == $this->length()){
                    return $this->removeLast();
                }
                else{
                    $cur = $this->head;
                    $c =0;
                    while ($cur != null){
                        if ($index == $c){
                            $cur = $cur->next;
                            $cur->previous->next = $cur->next;
                            $cur->next->previous = $cur->previous;
                            $cur = null;
                            $this->count--;
                        }
                        $c++;
                        $cur = $cur->next;
                    }
                }
            }
        }
    }
    //O(1)
    public function clear()
    {
        $this->tail = null;
        $this->head = null;
    }
}

File: LinkedList/C#/LinkedList/LinkedList/DoublyLinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LinkedList
{
    class DoublyLinkedList<T>
    {
        #region Properties
        public int Count { private set; get; }
        public DoublyLinkedListNode<T> Head { private set; get; }
        public T HeadValue
        {
            get
            {
                if (this.Head != null)
                {
                    return this.Head.Value;
                }
                else
                {
                    return default(T);
                }
            }
        }
        public DoublyLinkedListNode<T> Tail { private set; get; }
        public T TailValue
        {
            get
            {
                if (this.Tail != null)
                {
                    return this.Tail.Value;
                }
                else
                {
                    return default(T);
                }
            }
        }
        #endregion
        #region Constructors
        public DoublyLinkedList()
        {
            this.Count = 0;
            this.Head = null;
            this.Tail = null;
        }
        #endregion
        #region Add
        //O(1)
        public void AddFirst(DoublyLinkedListNode<T> node)
        {
            if (this.Head == null)
            {
                this.Head = node;
                this.Tail = node;
            }
            else
            {
                node.Next = this.Head;
                this.Head = node;
            }
            this.Count++;
        }
        //O(1)
        public void AddFirst(T vlaue)
        {
            DoublyLinkedListNode<T> node = new DoublyLinkedListNode<T>(vlaue);
            if (this.Head == null)
            {
                this.Head = node;
                this.Tail = node;
            }
            else
            {
                node.Next = this.Head;
                this.Head = node;
            }
            this.Count++;
        }

        //O(1)
        public void AddLast(DoublyLinkedListNode<T> node)
        {
            if (this.Tail == null)
            {
                this.Head = node;
                this.Tail = node;
            }
            else
            {
                this.Tail.Next = node;
                node.Previous = this.Tail;
                this.Tail = node;
            }
            this.Count++;
        }
        //O(1)
        public void AddLast(T vlaue)
        {
            DoublyLinkedListNode<T> node = new DoublyLinkedListNode<T>(vlaue);
            if (this.Tail == null)
            {
                this.Head = node;
                this.Tail = node;
            }
            else
            {
                this.Tail.Next = node;
                node.Previous = this.Tail;
                this.Tail = node;
            }
            this.Count++;
        }

        //O(n)
        public void Add(DoublyLinkedListNode<T> node, int index)
        {
            DoublyLinkedListNode<T> cur = this.Head;
            int c = 0;
            while (cur != null)
            {
                if (c == index)
                {
                    if (c == 0)
                    {
                        this.AddFirst(node);
                        cur = null;
                    }
                    else if (c == this.Count)
                    {
                        this.AddLast(node);
                        cur = null;
                    }
                    else
                    {
                        node.Previous = cur.Previous;
                        cur.Previous = node;
                        node.Next = cur;
                        cur = null;
                    }
                    this.Count++;
                }
                cur = cur.Next;
                c++;
            }
        }

        public void Add(T vlaue, int index)
        {
            DoublyLinkedListNode<T> node = new DoublyLinkedListNode<T>(vlaue);
            DoublyLinkedListNode<T> cur = this.Head;
            int c = 0;
            while (cur != null)
            {
                if (c == index)
                {
                    if (c == 0)
                    {
                        this.AddFirst(node);
                        cur = null;
                    }
                    else if (c == this.Count)
                    {
                        this.AddLast(node);
                        cur = null;
                    }
                    else
                    {
                        node.Previous = cur.Previous;
                        cur.Previous = node;
                        node.Next = cur;
                        cur = null;
                    }
                    this.Count++;
                }
                cur = cur.Next;
                c++;
            }
        }
        #endregion
        #region Remove
        //O(1)
        public void RemoveFirst()
        {
            if (this.Head != null)
            {
                var temp = this.Head;
                this.Head = this.Head.Next;
                temp.Next = null;
                this.Count--;
            }

        }
        //O(1)
        public void RemoveLast()
        {
            if (this.Tail != null)
            {  
                this.Tail = this.Tail.Previous;
                this.Tail.Next = null;
                this.Count--;
            }
        }

        //O(n)
        public void Remove(int index)
        {
            DoublyLinkedListNode<T> cur = this.Head;
            int c = 0;
            while (cur != null)
            {
                if (c == index)
                {
                    if (c == 0)
                    {
                        this.RemoveFirst();
                        cur = null;
                    }
                    else if (c == this.Count)
                    {
                        this.RemoveLast();
                        cur = null;
                    }
                    else
                    {
                        cur.Previous.Next = cur.Next;
                        cur.Next = null;
                        cur.Previous = null;
                        this.Count--;
                    }
                }
                cur = cur.Next;
                c++;
            }
        }
        #endregion

        #region Get
        //O(n)
        public T Get(int index)
        {
            int c = 0;
            DoublyLinkedListNode<T> cur = this.Head;
            bool found = false;
            while (cur != null)
            {
                if (c == index)
                {
                    found = true;
                    break;
                }
                cur = cur.Next;
                c++;
            }
            if (found)
            {
                return cur.Value;
            }
            else
            {
                return default(T);
            }
        }

        //O(n)
        public DoublyLinkedListNode<T> GetNode(int index)
        {
            int c = 0;
            DoublyLinkedListNode<T> cur = this.Head;
            while (cur != null)
            {
                if (c == index)
                {
                    break;
                }
                cur = cur.Next;
                c++;
            }
            return cur;
        }
        #endregion
        #region Clear
        public void Clear()
        {
            this.Head = null;
            this.Tail = null;
            this.Count = 0;
        }
        #endregion
    }
}
